### PR TITLE
JSONView collapse on click

### PR DESCRIPTION
**Pull request by mbecker20**
solves issue #57

### CODE REVIEW

**Comment by orbit-stabilizer** [comment]
I like the implementation - though I see one thing to note and one problem. The first is that this doesn't completely solve #57 because the second part is still needing to be done. And second, editing the textarea is difficult with this because clicking inside the textarea immediately shrinks and unshrinks it. 

Instead, maybe a button or something else outside the textarea to click might be better. If we do end up going that route, we can have max-height be 0% since we don't need to click on the textarea to open it again.

File: src/components/JSONView.tsx [hunk 1]
```diff
@@ -48,6 +48,11 @@ export const JSONView: FC<RudolfStore & { dispatch: CustomDispatch }> = ({
         })
     }
   }, [dispatch, justifications, tree])
+
+  const [open, setOpen] = useState(true)
+  function onClick() {
```
Comment (orbit-stabilizer) [inline src/components/JSONView.tsx:53]: It's not required to make a named function for a small task.

File: src/components/JSONView.tsx [hunk 2]
```diff
@@ -58,8 +63,9 @@ export const JSONView: FC<RudolfStore & { dispatch: CustomDispatch }> = ({
         minWidth: '100%',
         position: 'fixed',
         bottom: 0,
-        maxHeight: '50%',
+        maxHeight: open ? '50%' : '2%',
       }}
+      onClick={onClick}
```
Comment (orbit-stabilizer) [inline src/components/JSONView.tsx:68]: Here, we can instead have `onClick={() => setOpen(!open}`.

**Comment by mbecker20** [comment]
the last three commits ([1](https://github.com/McTano/Rudolf/pull/61/commits/e23aedae55d0de6d8275ffa7f1f9d8163d5c6769), [2](https://github.com/McTano/Rudolf/pull/61/commits/77172dee9dea0b70de015d68b4ebbeda06e07bad), [3](https://github.com/McTano/Rudolf/pull/61/commits/27270833c912b575073be79b20bbd132f20fffee)) solve the first part of issue 57 with a button, as you suggested

**Comment by orbit-stabilizer** [comment]
It looks really nice visually! It looks good to me - let's see what @McTano has to say.

**Comment by mbecker20** [comment]
As for the second part of 57, I got close to the required functionality by using a useState initialized to JSON.stringify(...). The value of the textarea becomes the return value of the useState. Then I think onChange just needs to dispatch the new value to the store. But now to update the value in JSONView, we would need to use the setValue func returned by the useState any time the state is updated outside of JSONView. Does this sound about right?

**Comment by McTano** [comment]
> As for the second part of 57, I got close to the required functionality by using a useState initialized to JSON.stringify(...). The value of the textarea becomes the return value of the useState. Then I think onChange just needs to dispatch the new value to the store. But now to update the value in JSONView, we would need to use the setValue func returned by the useState any time the state is updated outside of JSONView. Does this sound about right?


Yes, but it's a bit simpler then that.
Adding another `useState` is unnecessary.
The value of the text area is computed from the component's props, so there's no need to manually update it locally. If the store gets updated, the props will change, and the component will automatically update through React magic.
The second part of your answer is right, though. The way to do this would be to add an action to the rudolf store which takes JSON representing the app state and updates the store, then dispatch it in the `onChange` handler.

However, after thinking about it, I no longer think this is a good idea, since the user would have to update basically the whole app state at once to avoid invalid creating an invalid state in between (and potentially crashing the component).
If we want to be able to restore the component from saved state, it would be better to do it a different way.

**Comment by McTano** [comment]
The current changes look good. Don't worry about the update feature.
You can go ahead and merge it.

**Comment by McTano** [comment]
I tried `JSONView.styles.ts` and it broke, so instead of messing with the file extension I went with `JSONView_styles.ts`, which works.

**Comment by McTano** [comment]
I'm gonna go ahead and merge this so I can resolve my checker's conflicts with master and get the sort of working checker deployed.

**Pull request merged**